### PR TITLE
netebpfext tracing changes

### LINKED ISSUE #1465: Add / improve tracing for ebpf-for-windows
### Describe the feature you'd like supported

This issue to track tracing changes which are needed for better debuggability. This includes 2 parts:

1. Identify areas / flows where any tracing is missing and add required traces.
2. Improve / enrich any existing traces if needed.

### Proposed solution

1. Identify areas / flows where any tracing is missing and add required traces.
2. Improve / enrich any existing traces if needed.

### Additional context

_No response_

### LINKED ISSUE #2218: Always on minimal tracing for ebpf-for-windows
This issue is to track any changes that need to be done to enable always-on tracing for ebpf-for-windows.

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes #1465
Fixes #2218

## Description

This PR enhances tracing for SOCK_ADDR classify functions.
1. Traces now include source IP / port, destination IP / port, redirected IP / port.
2. BLOCK and REDIRECTED connection are traced at INFO level. ALLOW is traced at VERBOSE level.

## Testing

Existing CICD

## Documentation

NA

### CODE REVIEW

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 1]
```diff
@@ -985,6 +985,71 @@ net_ebpf_ext_connect_redirect_filter_change_notify(
     NET_EBPF_EXT_RETURN_NTSTATUS(status);
 }
 
+#define DECLARE_SOCK_ADDR_CLASSIFY_LOG_FUNCTION(family)                 \
+    static void _net_ebpf_ext_log_sock_addr_classify_v##family##(       \
+        _In_z_ const char* message,                                     \
+        uint64_t transport_endpoint_handle,                             \
+        _In_ const bpf_sock_addr_t* original_context,                   \
+        _In_opt_ const bpf_sock_addr_t* redirected_context,             \
+        uint32_t verdict)                                               \
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_sock_addr.c:994]: Seems like the `verdict` type should be an enum type not `uint32_t`

File: netebpfext/net_ebpf_ext_tracelog.h [hunk 2]
```diff
@@ -209,16 +209,90 @@ net_ebpf_ext_trace_terminate();
...
+        TraceLoggingLevel(trace_level),                                                                        \
+        TraceLoggingKeyword(NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR),                                          \
+        TraceLoggingString(message, "Message"),                                                                \
+        TraceLoggingUInt64((handle), "TransportEndpointHandle"),                                               \
+        TraceLoggingUInt64((protocol), "Protocol"),                                                            \
+        TraceLoggingIPv4Address((source_ip), "src_ip"),                                                        \
+        TraceLoggingUInt16((source_port), "src_port"),                                                         \
+        TraceLoggingIPv4Address((destination_ip), "dst_ip"),                                                   \
+        TraceLoggingUInt16((destination_port), "dst_port"),                                                    \
+        TraceLoggingUInt64((verdict), "Verdict"));
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_tracelog.h:226]: why 64-bit instead of (say) 32-bit?

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 1 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c:994]: I think the verdict not an enum in Linux also -- `BPF_SOCK_ADDR_VERDICT_PROCEED`, `BPF_SOCK_ADDR_VERDICT_REJECT`

File: netebpfext/net_ebpf_ext_tracelog.h (hunk 2 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_tracelog.h]: Changed to `TraceLoggingUInt32`

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 3]
```diff
@@ -985,6 +985,71 @@ net_ebpf_ext_connect_redirect_filter_change_notify(
     NET_EBPF_EXT_RETURN_NTSTATUS(status);
 }
 
+#define DECLARE_SOCK_ADDR_CLASSIFY_LOG_FUNCTION(family)                 \
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:988]: nit, this is `DEFINE` not `DECLARE`.

File: netebpfext/net_ebpf_ext_tracelog.h [hunk 4]
```diff
@@ -209,16 +209,90 @@ net_ebpf_ext_trace_terminate();
...
-    TraceLoggingWrite(                                                                                          \
-        net_ebpf_ext_tracelog_provider,                                                                         \
-        NET_EBPF_EXT_TRACELOG_EVENT_GENERIC_MESSAGE,                                                            \
-        TraceLoggingLevel(trace_level),                                                                         \
-        TraceLoggingKeyword((keyword)),                                                                         \
-        TraceLoggingString(message, "Message"),                                                                 \
-        TraceLoggingUInt64((handle), "TransportEndpointHandle"),                                                \
-        TraceLoggingUInt64((protocol), "Protocol"),                                                             \
-        TraceLoggingUInt64((redirect), "Redirected"),                                                           \
+#define NET_EBPF_EXT_LOG_SOCK_ADDR_CLASSIFY_IPV4(                                                              \
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_tracelog.h:212]: move these to `net_ebpf_ext_sock_addr.h`.

**Review by shankarseal — APPROVED** [review]
Approved modulo the changes suggested.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 1 above)
Comment (dthaler) [reply netebpfext/net_ebpf_ext_sock_addr.c:994]: Are you updating this in this PR?

Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c:994]: The `verdict` is return value from the ebpf program. it is an int right now (and not changed in this PR). I think we can update it in a separate PR? Also, if we are planning to move to enums, we should ideally change it for all the program types.

File: netebpfext/net_ebpf_ext_tracelog.h (hunk 4 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_tracelog.h:212]: Shouldn't all tracing macros be in one file, even if some of them are sock_addr specfic?

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 3 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c]: fixed.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 1 above)
Comment (dv-msft) [reply netebpfext/net_ebpf_ext_sock_addr.c:994]: I would also prefer changing it for all program types at the same time. I'm ok with moving the 'enum' update to a separate PR.

File: netebpfext/net_ebpf_ext_tracelog.h (hunk 4 above)
Comment (dv-msft) [reply netebpfext/net_ebpf_ext_tracelog.h:212]: Given that these macros are not dependent on specific header files, I'd personally prefer all tracing macros to be in one file as well.

**Review by dv-msft — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: netebpfext/net_ebpf_ext_tracelog.h (hunk 4 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_tracelog.h:212]: Disagree. Module specific tracing should be local to the module. I will fix this.

**Pull request merged**